cachi2/core/package_managers/yarn/locators.py

```diff
@@ -0,0 +1,236 @@
+        """Parse a Locator from a string."""
+        parts = locator.split("@")
+
+        # aliased packages start with @
+        if parts[0] == "":
+            return NpmLocator(name="@" + parts[1], version=parts[2])
+
+        return NpmLocator(name=parts[0], version=parts[1])
+
+    def __str__(self) -> str:
+        return f"{self.name}@{self.version}"
+
+
+class PatchLocator(NamedTuple):
+    """Locator that handles patched dependencies.
+
+    Sample locator string:
+    dependency-name@patch:patched-dependency@npm%3A0.0.1#./path/to/patch::version=1.0.0&hash=abc123&locator=package-name%40workspace%3A
+    dependency-name@patch:patched-dependency@npm%3A0.0.1#built-in<compat/patch-name>::version=1.0.0&hash=abc123 (built-in patch)
+    """
+
+    name: str
+    patches: "Locator"
+    patch: str
+
+    @property
+    def protocol(self) -> str:
+        """The protocol present in this locator."""
+        return "patch"
+
+    @classmethod
+    def from_string(cls, locator: str) -> "PatchLocator":
+        """Parse a Locator from a string."""
+        patches = NpmLocator(name="another-dep", version="0.0.1")
+        return PatchLocator(
+            name="mock",
+            patches=patches,
+            patch="./path/to/patch::version=1.0.0&hash=abc123&locator=package-name%40workspace%3A",
+        )
+
+    def __str__(self) -> str:
+        return f"{self.name}@patch:{self.patches}#{self.patch}"
+
+
+class WorkspaceLocator(NamedTuple):
+    """Locator that handles workspace dependencies.
+
+    Sample locator string:
+    dependency-name@workspace:path/to/dir
+    """
+
+    name: str
+    path: str
+
+    @property
+    def protocol(self) -> str:
+        """The protocol present in this locator."""
+        return "workspace"
+
+    @classmethod
+    def from_string(cls, locator: str) -> "WorkspaceLocator":
+        """Parse a Locator from a string."""
+        parts = locator.split("@workspace:")
+
+        return WorkspaceLocator(parts[0], parts[1])
+
+    def __str__(self) -> str:
+        return f"{self.name}@worskpace:{self.path}"
+
+
+def parse_locator(locator: str) -> Locator:
+    """Parse a Locator object based on a locator string from the 'yarn info' command."""
+    if "@file:" in locator or "@link:" in locator or "@portal:" in locator:
+        return FileLocator.from_string(locator)
+
+    if "#commit=" in locator:
+        return GitLocator.from_string(locator)
+
+    if "@https://" in locator:
+        return HttpsLocator.from_string(locator)
+
+    if "@npm:" in locator:
+        return NpmLocator.from_string(locator)
+
+    if "@patch:" in locator:
+        return PatchLocator.from_string(locator)
+
+    if "@workspace:" in locator:
+        return WorkspaceLocator.from_string(locator)
+
+    raise PackageRejected(
+        reason=f'Locator "{locator}" couldn\'t be parsed.',
+        solution="Cachi2 only supports the default Yarn protocols: https://v3.yarnpkg.com/features/protocols.",
+    )
```

Comment:
All the parsing methods and parse_locator should be TODO or removed as well IMO

I would prefer removed to avoid leading the implementer down the wrong path (we really, really should check how yarnberry parses them)